File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
from collections import deque

N, E = map(int, sys.stdin.readline().strip().split())

def bfs(temp):
  global arr, visited
  q = deque(arr[temp])

  while q:
    temp = q.popleft()
    if not visited[temp]:
      visited[temp] = True
      for i in range(len(arr[temp])):
        q.append(arr[temp][i])
      

arr = [[] for _ in range(N + 1)]
visited = [False] * (N + 1)
count = 0

for i in range(E):
  a, b = map(int, sys.stdin.readline().strip().split())
  
  # 무방향이라
  
  arr[a].append(b)
  arr[b].append(a)
  
for i in range(1, N):
  if not visited[i]:
    visited[i] = True
    bfs(i)
    count += 1
  
visited_cnt = 0  
for i in range(1, len(visited)):
  if not visited[i]: visited_cnt += 1
print(count + visited_cnt)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;


public class Main {

    static List<Integer>[] graph;
    static boolean[] check;
    // 정답 개수
    static int count = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        String[] s = bf.readLine().split(" ");
        int N = Integer.parseInt(s[0]);
        int M = Integer.parseInt(s[1]);

        graph = new List[N+1];

        for(int i=0; i<N+1; i++){
            graph[i] = new ArrayList<>();
        }

        check = new boolean[N+1];

        for(int i=0; i<M; i++){
            s = bf.readLine().split(" ");
            int start = Integer.parseInt(s[0]);
            int end = Integer.parseInt(s[1]);

            graph[start].add(end);
            graph[end].add(start);
        }

        for(int i=1; i<N+1; i++){
            if(!check[i]){
                count++;
                bfs(i);
            }
        }

        System.out.println(count);
    }

    public static void bfs(int i){
        Queue<Integer> q = new LinkedList<>();
        q.add(i);
        check[i] = true;

        while(q.size() != 0){
            int x = q.poll();
            for(int j=0; j<graph[x].size(); j++){
                int temp = graph[x].get(j);
                if(!check[temp]){
                    q.add(temp);
                    check[temp] = true;
                }
            }
        }
    }
}